Loading state for Button component

greglo opened the pull request:
#### Fixes #198

### Button now accepts optional `loading` prop

Default: `false`

If `loading` is set to true, the button will display a centered loading spinner instead of
the button contents. The size of the button will not change based on the
value of the `loading` prop.

### Disable on loading?

Right now the button gets disabled if `loading` is set to true. At least for my own use-case that is the desired behavior, and I struggled to cook up a compelling counterexample in my head, but can believe someone might object. Thoughts?

### Screenshots

Default:
![image](https://cloud.githubusercontent.com/assets/2872321/21089385/81b7b946-bfec-11e6-8e95-507f889832e7.png)

Primary
![image](https://cloud.githubusercontent.com/assets/2872321/21089394/8e7d763e-bfec-11e6-9dd5-2e037e57a308.png)

Primary, minimal
![image](https://cloud.githubusercontent.com/assets/2872321/21089408/a53efbd6-bfec-11e6-9d79-9c5f762af747.png)


Dark
![image](https://cloud.githubusercontent.com/assets/2872321/21089414/b543525c-bfec-11e6-8f07-4c388ae3d0bb.png)

Dark, primary
![image](https://cloud.githubusercontent.com/assets/2872321/21089421/bd576280-bfec-11e6-9693-020bf991feb8.png)

Dark, primary, minimal
![image](https://cloud.githubusercontent.com/assets/2872321/21089435/db062fbe-bfec-11e6-9c25-be0fe8e86f1a.png)

adidahiya commented:
@greglo can you please enable Circle CI on your fork?

llorca commented:
Wow thanks for contributing this awesome little feature! I'll do a thorough review today. Looks like you even nailed the visual design (spinner coloring), major props

llorca commented on packages/core/examples/buttonsExample.tsx:
```diff
@@ -16,16 +16,18 @@ import { IntentSelect } from "./common/intentSelect";
 export interface IButtonsExampleState {
     disabled?: boolean;
     intent?: Intent;
+    isLoading?: boolean;
```
I'd go for `loading` -- it's simpler and more consistent with other `boolean` usage in this interface

llorca commented on packages/core/src/components/button/_button.scss:
```diff
@@ -49,6 +49,7 @@ Styleguide components.button.css
 .pt-button {
   @include pt-button-base();
   @include pt-button-height($pt-button-height);
+  position: relative;
```
Is that fine? I wonder if it could be breaking in some cases. Would need confirmation from Gilad/Chris/Adi

giladgray replied:
👎  until we get a report on testing this in some real applications.

llorca commented on packages/core/src/components/button/_button.scss:
```diff
@@ -260,6 +261,17 @@ a.pt-button {
   @extend .pt-button:disabled;
 }
 
+.pt-button-loading-spinner {
```
I feel like it'd be neat to select this spinner via composition instead of custom class: `.pt-button .pt-spinner`

greglo replied:
Note that the `pt-button-loading-spinner` (perhaps poorly named) is not the `.pt-spinner`. It is the extra element discussed below

adidahiya replied:
@llorca as a _very_ broad/rough rule, I would prefer custom classes because they reduce CSS specificity

@greglo I think the name looks good

adidahiya replied:
Wait, nvm, I misread that. If this corresponds to the spinner container element, go for `.pt-button-spinner-container`

llorca replied:
I get that. With Adi here, if it's the spinner container, `.pt-button-spinner-container` sounds good

llorca commented on packages/core/src/components/button/_common.scss:
```diff
@@ -154,6 +154,10 @@ $non-default-state-selectors: ":last-child, :hover, :active, .pt-active, [class*
   }
+
+  .pt-button-loading-spinner .pt-spinner-head {
+    stroke: $white;
```
Yes 👏 👏

llorca commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
 }
 
+function isButtonDisabled(props: IButtonProps) {
+    return props.disabled === true || props.isLoading === true;
```
To answer your question, yes, I'd also expect a "loading" button to be disabled. Let's go with that 👍

llorca commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+
+function maybeRenderSpinner(isLoading: boolean) {
+    return isLoading != null && isLoading
+      ? <span className="pt-button-loading-spinner"><Spinner className="pt-small" /></span>
```
See the other comment above -- we usually prefer composition, so I believe it's possible to throw a spinner without the custom class here. There may be some drawbacks I'm not aware of

greglo replied:
I'm not sure we can. I want the parent here to take up the full area of the button so I can center the spinner using `flex`. It may be possible without the extra element... @giladgray @adidahiya ?

llorca commented on packages/core/src/components/button/_common.scss:
```diff
@@ -184,6 +188,10 @@ $non-default-state-selectors: ":last-child, :hover, :active, .pt-active, [class*
   }
+
+  .pt-button-loading-spinner .pt-spinner-head {
+    stroke: $pt-dark-text-color;
```
For these values, we may be able to reuse some color variables from https://github.com/palantir/blueprint/blob/master/packages/core/src/components/spinner/_spinner.scss

llorca replied:
Whoops nevermind, that's the `head` not the `track`

greglo commented:
@adidahiya @giladgray Would be great to get your input on the use of `position: relative`

llorca commented on packages/core/src/components/button/_button.scss:
```diff
@@ -260,6 +261,17 @@ a.pt-button {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  position: absolute;
```
Why do you need both flex and absolute positioning?

greglo replied:
My logic here (perhaps faulty) is to absolutely position the container to take the loading spinner out of flow, and then use flex so its child (the spinner) can be trivially positioned in the middle of this component.

adidahiya replied:
yeah that makes sense @greglo. `display: flex` is a little unintuitive because it primarily sets up positioning for its children, not itself -- other `display` values do the opposite.

adidahiya commented:
It's unfortunate that @blueprint-bot doesn't post previews for PRs from forks. I imagine it has something to do with this line in circle config: https://github.com/palantir/blueprint/blob/497a03eab4420e48db0865407995895aa25726fa/circle.yml#L31

greglo commented:
@llorca I believe I have addressed all your comments

@adidahiya Just waiting for comment on the addition of `position: relative` at the root button level.

adidahiya commented:
I can't say for certain whether `.pt-button { position: relative }` is breaking. Can you test it out on some larger applications with the devtools CSS editor @greglo?

giladgray commented:
yes @adidahiya's suggestion above is the best way to validate that `position` is safe to add (i think it should be). @greglo think you could tackle that this week?

giladgray commented on packages/core/src/components/button/_common.scss:
```diff
@@ -184,6 +188,10 @@ $non-default-state-selectors: ":last-child, :hover, :active, .pt-active, [class*
     color: $dark-button-color-disabled;
   }
+
+  .pt-button-spinner-container .pt-spinner-head {
```
i suspect `.pt-button-spinner-container` is unnecessary in this selector and can be safely removed. 

will there ever be another `.pt-spinner-head` inside a button? if so, would you ever want that head to be a _different_ color?

greglo replied:
Well we don't know what content the user puts in the `Button` right, so they might put in a `Spinner intent={Intent.Primary}` and be sad that the stroke isn't what they expect?

Perhaps they shouldn't be doing that, but I'd argue that its better to just faithfully render `children` without overriding its styles

giladgray replied:
disagree, i don't think we should support red spinners inside blue buttons, for instance. things are always white inside an intent.

greglo replied:
_shrug_ I defer to you, but I would not expect this as a consumer.

llorca replied:
Indeed we don't have much choice for non-minimal intent buttons, the spinner needs to be white

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -13,20 +13,26 @@ import * as React from "react";
     /** Name of icon (the part after `pt-icon-`) to add to button. */
     rightIconName?: string;
+
+    /** Whether the button's action is currently loading (will disable the button) */
```
not the best explanation of what the prop does. i like what you have in the PR description about showing a spinner but preserving the button width.

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
 }
 
+function isButtonDisabled(props: IButtonProps) {
+    return props.disabled === true || props.loading === true;
```
safe to drop `=== true` here, we know they're booleans.

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+}
+
+function buttonContentsVisibility(loading: boolean) {
+    return {visibility: loading ? "hidden" : "visible"};
```
oof this line makes this whole thing much trickier to use without JS. i know we're not targeting a CSS API but (until I saw this) all it required was adding the spinner in `.pt-button-spinner-container`. 
would be great to refactor this to a simple selector by adding `.pt-loading` modifier to `.pt-button`

giladgray replied:
i see why this is tricky to work around and the `span` may be necessary, but it's the one thing in this PR that makes me most uncomfortable.

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+
+function maybeRenderSpinner(loading: boolean) {
+    return loading != null && loading
+      ? <span className="pt-button-spinner-container"><Spinner className="pt-small" /></span>
```
is there any way to re-use the `Spinner`'s root element as the `pt-button-spinner-container`? so we don't need to add yet another element.

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+}
+
+function maybeRenderSpinner(loading: boolean) {
+    return loading != null && loading
```
simpy `loading` is enough, we know it's a boolean and that includes the is-defined check.

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +88,27 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+}
+
+function buttonContentsVisibility(loading: boolean) {
+    return {visibility: loading ? "hidden" : "visible"};
```
spaces inside object literal braces please. `{ visiblity ... }`

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -13,20 +13,27 @@ import * as React from "react";
     rightIconName?: string;
+
+    /** If set to true, the button will display a centered loading spinner instead of the button contents. The size of
+     * the button is not affected by the value of this prop. */
```
split over multiple lines. also add default value.

```
/**
 * line 1
 * line 2
 * @default false 
 */
```

giladgray commented:
re @adidahiya's comment about @blueprint-bot not positing comments: #347 removes that `owner` line to make this possible.

greglo commented:
@giladgray I am going to be able to get rid of some elements, but don't think I can avoid adding at least one new span. The reason is I can't set the `{text}` to `visibility: hidden` without wrapping it in an element, unless you know how to hide the child text of a Button directly? I wasn't able to find a suitable way

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +84,24 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
         Classes.BUTTON,
-        { [Classes.DISABLED]: props.disabled },
+        { [Classes.DISABLED]: isButtonDisabled(props) },
+        { [Classes.LOADING]: props.loading },
```
merge with object above. two props, one object!

llorca replied:
I think he means:
```
Classes.BUTTON, {
    [Classes.DISABLED]: isButtonDisabled(props),
    [Classes.LOADING]: props.loading,
},
```

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -35,7 +42,8 @@ export class Button extends React.Component<React.HTMLProps<HTMLButtonElement> &
             >
-                {text}
+                {maybeRenderSpinner(loading)}
+                <span>{text}</span>
```
`text` is optional; can we render this conditionally somehow?

giladgray commented on packages/core/src/components/button/_button.scss:
```diff
@@ -68,6 +68,21 @@ Styleguide components.button.css
     }
   }
 
+  &.pt-loading {
```
👍 great use of modifier.

giladgray commented on packages/core/src/components/button/_button.scss:
```diff
@@ -68,6 +68,21 @@ Styleguide components.button.css
+  &.pt-loading {
+    display: flex;
+    align-items: center;
+    justify-content: center;
```
flex seems a little over-the-top here when we're going to hide all the real content and absolutely position the spinner. might `display: flex` also end up changing the size of certain complex buttons?

giladgray replied:
how about doing the `translate(-50%, -50%)` trick on the spinner itself?

greglo replied:
Note that using `position: absolute` without setting `top`, `right`, etc. means it gets positioned where it otherwise would have (hence `display: flex`). I can play around with `translate` in a separate commit in a minute

adidahiya replied:
IMO flex is not over the top and I generally prefer it over `position: absolute` approaches

giladgray replied:
really want to see a preview so I can play with the DOM...

adidahiya commented on packages/core/examples/buttonsExample.tsx:
```diff
@@ -48,12 +51,24 @@ export class ButtonsExample extends BaseExample<IButtonsExampleState> {
-                    intent={this.state.intent}
-                    onClick={this.beginWiggling}
-                    text="Click to wiggle"
+                  {...removeNonHTMLProps(this.state, INVALID_HTML_PROPS)}
```
use 4-space indentation please, here and below (I suggest configuring EditorConfig)

adidahiya commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,24 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+    return props.disabled || props.loading;
+}
+
+function renderSpinner(loading: boolean) {
```
`maybeRenderSpinner`

llorca commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -73,13 +87,30 @@ export const AnchorButtonFactory = React.createFactory(AnchorButton);
+
+function maybeRenderSpinner(loading: boolean) {
+    return loading
+      ? <Spinner className="pt-small pt-button-spinner" />
```
Apologies if I asked already, but instead of `.pt-button-spinner`, can't we use `.pt-button.pt-loading .pt-spinner`?

giladgray replied:
yeah seems reasonable. we can target this element with a descendant selector. would also support custom spinner usage (like if you want to render text and spinner).

llorca replied:
Cool, that makes for better composition

llorca commented on packages/core/src/components/button/_common.scss:
```diff
@@ -154,6 +154,10 @@ $non-default-state-selectors: ":last-child, :hover, :active, .pt-active, [class*
     background-image: none;
   }
+
+  .pt-button-spinner .pt-spinner-head {
```
What about `.pt-button.pt-loading .pt-spinner` instead of `.pt-button-spinner`?

greglo commented:
Is renaming those selectors (although again, note, they are not semantically equivalent) the last change?

pkwi commented:
@greglo just noticed that your dark spinner is using a different colour from the one in docs. Where there any contrast issues with our standard colouring?

greglo commented:
@pkwi Are you referring to use of `pt-dark-text-color` instead of `pt-dark-button-color-disabled`? (https://github.com/palantir/blueprint/pull/349/files#diff-a4dfb8855074f380465eaa9d35f20bb0R193).

I seem to remember that using the darker color was a bit hard to see for me, but defer to your

llorca commented on packages/core/src/components/button/_common.scss:
```diff
@@ -184,6 +188,10 @@ $non-default-state-selectors: ":last-child, :hover, :active, .pt-active, [class*
   }
+
+  .pt-button-spinner .pt-spinner-head {
+    stroke: $pt-dark-text-color;
```
@greglo I think that's the one Piotr is talking about. We could use `$dark-spinner-head-color` to be consistent with spinners. Don't worry about it though, I'll take care of that as part of https://github.com/palantir/blueprint/issues/369

giladgray replied:
#369 was implemented in the latest release. @greglo please update these colors accordingly.

dmackerman commented:
One request I'd have for this is the ability to keep the `text` property visible, because sometimes it's nice to inform the user of what "process" is happening based on the button.

giladgray commented on packages/core/examples/buttonsExample.tsx:
```diff
@@ -36,7 +38,8 @@ export class ButtonsExample extends BaseExample<IButtonsExampleState> {
 
-    private timeoutId: number;
+    private wiggleTimeoutId: number;
+    private loadingTimeoutId: number;
```
(done)

giladgray commented on packages/core/examples/buttonsExample.tsx:
```diff
@@ -57,6 +65,18 @@ export class ButtonsExample extends BaseExample<IButtonsExampleState> {
+                    intent={this.state.intent}
+                    loading={this.state.loading}
+                    onClick={this.beginLoading}
+                    text="Click to build"
```
"Click to load" ?

greglo replied:
Done

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -13,20 +13,30 @@ import * as React from "react";
     rightIconName?: string;
+
+    /**
+     * If set to true, the button will display a centered loading spinner instead of the button contents.
```
"instead of its contents" ?

greglo replied:
Done

giladgray commented on packages/core/src/components/button/buttons.tsx:
```diff
@@ -13,20 +13,30 @@ import * as React from "react";
+
+    /**
+     * If set to true, the button will display a centered loading spinner instead of the button contents.
+     * The size of the button is not affected by the value of this prop.
```
"The width"

greglo replied:
I can change, although work noting that it is significant that the height doesn't change too :)

llorca commented:
@dmackerman why both a spinner and a text? If text is required, a simple `disabled` prop should do the trick. One of the good things about our current approach is that the button size doesn't change when the loading or disabled modifier is applied

giladgray commented:
[docs preview from latest build](https://19-76082203-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/docs/dist/index.html)

giladgray commented:
[docs preview from latest build](https://21-76082203-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/docs/dist/index.html) is __fixed__ and looks __:fire:__

giladgray commented on circle.yml:
```diff
@@ -27,7 +27,7 @@ dependencies:
   post:
     - lerna bootstrap
-    - gulp build
+    - npm run build:gulp
```
this fixes #451
